include/swift/AST/DiagnosticConsumer.h

```diff
@@ -140,10 +140,19 @@ class FileSpecificDiagnosticConsumer : public DiagnosticConsumer {
   /// All consumers owned by this FileSpecificDiagnosticConsumer.
   const SmallVector<ConsumerPair, 4> SubConsumers;
 
-  /// The DiagnosticConsumer may be empty if those diagnostics are not to be
-  /// emitted.
-  using ConsumersOrderedByRangeEntry =
-    std::pair<CharSourceRange, DiagnosticConsumer *>;
+  // The commented-out consts are there because the data does not change
+  // but the swap method gets called on this structure.
+  struct ConsumerSpecificInformation {
```

Comment:
One consequence of putting the flag in with the mapping like this is that the same consumer might get multiple copies of the dummy error, one for each range (file) it's covering. We're not reusing consumers that way, so it doesn't matter, but I thought I'd bring it up and see what you thought.

(I was originally thinking it'd be a `SmallPtrSet<DiagnosticConsumer *, N>` that tracked all of the consumers that actually received non-note diagnostics. That *would* avoid the `const_cast`, but at the practical, run-time cost of an extra lookup.)

Reply:
I like the separation of concerns in your scheme; I never thought about reusing the consumers. I would rather put the flag in the consumer, and let a diagnostic consumer know if it had seen an error. But I got the impression that you disliked adding that functionality to the general consumer & I can see that. For now, I would prefer sticking with the current scheme.